### ISSUE TITLE
Implement itch.io as a new game service

### ISSUE THREAD

**Issue by neuromancer** [tkashkin/GameHub#65]
Reference implementations:

* https://github.com/itchio/itch 
* https://github.com/itchio/butler

**Comment by tkashkin** [comment]
It doesn't look like they have required functionality in their API. Also OAuth applications can only request profile (https://itch.io/docs/api/oauth#the-authorization-step/scopes).

**Comment by neuromancer** [comment]
Can you just use buttle to download and update games? (I know it is not the best solution..)

**Comment by neuromancer** [comment]
An interesting feature to consider from the itch.io client is [the use of a sandbox](https://itch.io/docs/itch/using/sandbox.html).

**Comment by impiaaa** [comment]
It looks like the official client uses a different API, used and sort-of-documented [here](https://github.com/itchio/go-itchio). Specific endpoints of interest might be /login, /profile/owned-keys, /games/&lt;gameid&gt;/uploads, /games/&lt;gameid&gt;/download-sessions, and /profile/game-sessions/summaries/&lt;gameid&gt;. Additionally, it seems like Playnite was able to get started by [contacting the itch devs directly](https://github.com/JosefNemec/Playnite/issues/107#issuecomment-445995670).

**Comment by hagabaka** [comment]
Hi, I have made some progress on this. https://github.com/hagabaka/GameHub/tree/itch

Currently I'm able to login with an API key, and fetch a list of games, although I haven't figured out why they're not showing up in the GameHub UI (it's able to print a list of their names in console).

This is much more complicated than other data sources, because `butler daemon` uses a flavor of JSON-RPC 2.0 which doesn't seem to be supported by json-rpc-glib, so I had to make my own implementation. The documentation for this API is currently here: http://docs.itch.ovh/butlerd/master

The code is missing a lot of error checking, and probably doing some things the wrong way. I'd appreciate any help and directions.

**Comment by tkashkin** [comment]
@hagabaka thanks. I have fetched your branch to https://github.com/tkashkin/GameHub/tree/itch.

I have added some error checks and improved `butler` detection. By default `butler` is not in `$PATH` if it was installed with itch desktop app. This case should be handled correctly now.

I can't actually test this since I don't have anything on itch.io. I assume `Fetch.ProfileOwnedKeys` should only return bought items.

Run GameHub with `--debug --verbose` arguments to see logs.

**Comment by hagabaka** [comment]
Thanks. I've merged the commits and will keep working on it. `Fetch.ProfileOwnedKeys` returns free games as well as purchased games. It *should* be printing a list of game names in the console even if you haven't bought anything...

**Comment by hagabaka** [comment]
I've added the functionality to list, install, and run itch.io games. Here are some of the remaining issues:

- Installing a game uses the first version of game offered by butler, which I think is the latest version for your platform. There is no functionality to install an alternative version though. Butler does provide the information needed to implement this, but I'm not sure how to implement the UI for it. `InstallDialog` seems interesting, but it takes a list of `Runnable.Installer` which I think need to have URLs, and butler doesn't provide them. So we would need to just display a dialog with a few choices, and allow one to be selected.

- There is no indication of "downloading" or "running". Butler provides these through JSON-RPC notifications. However they are not automatically associated with tasks. To properly handle them, it recommends opening dedicated connections for longer running tasks like installations and launches, which will only receive notifications for the tasks they start.

- Updating game information on demand isn't implemented yet, so the information is only loaded from butler when GameHub starts.

- Uninstalling a game isn't implemented yet.

- Running with compatibility tool isn't implemented. Currently the implementation uses Butler's `Launch` request to run games, which doesn't support this feature. I'm not sure what the data source needs to do to support this actually.

**Comment by tkashkin** [comment]
* Yes, `Runnable.Installer` should be implemented for itch.io games. It may require some changes but should be possible.
* `Downloader` and `Download`/`PausableDownload` are abstract classes. Probably there should be itch-specific `Download` implementation which will manage butler connection, listen to events and update download status.
* Is it possible to get game's executable path from butler? It should be possible to use compatibility layers in that case.

**Comment by tkashkin** [comment]
@hagabaka I have refactored `Runnable.Installer` in 38f7c2b.
Now there are `Installer`, `FileInstaller` and `DownloadableInstaller` base abstract classes.

Now GameHub should load installed free itch games in addition to owned games.

Also some GameHub features are implemented for itch games.

**Comment by fasterthanlime** [comment]
@hagabaka Using separate connections for tasks like launching/installing/etc. is a good idea indeed! Closing the connection will "cancel" these operations. This is the concept of "conversations" as implemented in https://github.com/itchio/node-butlerd

Here's an example usage of node-butlerd:

https://github.com/itchio/itch/blob/308582de070376fb017a4f2917e42a6d8116cdeb/src/main/reactors/updater.ts#L73-L109

Having a separate connection lets you route logging (the `Log` message) elsewhere, cancel separately, not mix up server requests (that should show dialogs, etc.). It's low-tech, but it works.

Like @aaronfranke said, I'm on semi-break, but happy to review butlerd client implementation before it goes live (I will be afk Sep 1-9 though).

**Comment by tkashkin** [comment]
@hagabaka I have refactored download status so it doesn't need workarounds like this.
There's now `ItchDownloader` which implements `Downloader` and has signals to update UI. That allows to integrate downloads into UI.
`Installer` is also implemented and now `InstallDialog` is used to select version to install.

**Comment by hagabaka** [comment]
@tkashkin Cool! That works great for me. 

I added handling of `ShellLaunch`, `HTMLLaunch`, and `URLLaunch` server requests. These are used when butler tries to open an HTML game. We're just using `xdg-open` right now, and some games don't work because they need to be served in a local HTTP server, but I don't think this would be a high priority issue for most people. I implemented this just so that GameHub isn't stuck in "running" when I click on such games, because butler is still waiting for the response.

It seems that the only main missing feature is running with compatibility layer. Butler doesn't have an API to tell us the name of the game executable. We could try to guess which file is the main executable, but there would be risk of running "uninstall.exe" etc. Butler has code to handle this (through [dash](https://github.com/itchio/dash)), but it doesn't tell us the result, and it filters by platform by default. I think we should ask if Butler API could provide this information, which will make launching with compatibility much easier to implement. It would be even better if Butler support that directly, and there is already an [issue](https://github.com/itchio/itch/issues/433) requesting it.

**Comment by hagabaka** [comment]
@tkashkin What other missing features or implements do you think we need to implement before merging this?

One possibility is updating games. Butler has an `CheckUpdates` API, which returns a list of `Upload`s that can be installed with similar calls as installing games. However GameHub doesn't seem to have UI for updating games. Should it automatically check for and install any updates before running a game?

**Comment by spinningD20** [comment]
Looking forward to using this!  What else needs to be done before we merge it in?  If you need me to test anything, I can give it a shot.

Thanks for the work on this, and for gamehub in general, I really like the project.

**Comment by tkashkin** [comment]
@spinningD20 testing would be nice.
If it works and it's stable I can probably find some free time and merge it.
Some kind of game updates UI for this and #323 would be nice, but meanwhile updates on game launch are fine.

**Comment by spinningD20** [comment]
```
[ERROR]  [GLib-GIO] Settings schema 'com.github.tkashkin.gamehub.paths' does not contain a key named 'itch-home'
Trace/breakpoint trap (core dumped)

```

There might be a previous settings file somewhere, but the above is what I'm getting when I do `ninja` in the build directory after doing a `meson` in the base directory, and attempt to run the compiled binary.

I'm on ArchLinux fwiw.

**Comment by tkashkin** [comment]
@spinningD20
You need to build package from this branch or run `sudo ninja install` to install updated settings schema.
Alternatively you can try to set schema directory before running compiled binary:
```bash
export GSETTINGS_SCHEMA_DIR="$APPDIR/usr/share/glib-2.0/schemas/:$GSETTINGS_SCHEMA_DIR"
```

**Comment by spinningD20** [comment]
`[wm@w-home build]$ sudo ninja install
[3/4] Installing files.
Installing subdir /home/wm/Downloads/GameHub/data/share/compat to /usr/share/com.github.tkashkin.gamehub/compat
Installing /home/wm/Downloads/GameHub/data/share/compat/dosbox/windowed.conf to /usr/share/com.github.tkashkin.gamehub/compat/dosbox
Installing data/com.github.tkashkin.gamehub.desktop to /usr/share/applications
Installing src/com.github.tkashkin.gamehub to /usr/bin
Installing /home/wm/Downloads/GameHub/data/icon/32.svg to /usr/share/icons/hicolor/32x32/apps
Installing /home/wm/Downloads/GameHub/data/icon/32.svg to /usr/share/icons/hicolor/32x32@2/apps
Installing /home/wm/Downloads/GameHub/data/icon/48.svg to /usr/share/icons/hicolor/48x48/apps
Installing /home/wm/Downloads/GameHub/data/icon/48.svg to /usr/share/icons/hicolor/48x48@2/apps
Installing /home/wm/Downloads/GameHub/data/icon/64.svg to /usr/share/icons/hicolor/64x64/apps
Installing /home/wm/Downloads/GameHub/data/icon/64.svg to /usr/share/icons/hicolor/64x64@2/apps
Installing /home/wm/Downloads/GameHub/data/icon/128.svg to /usr/share/icons/hicolor/128x128/apps
Installing /home/wm/Downloads/GameHub/data/icon/128.svg to /usr/share/icons/hicolor/128x128@2/apps
Installing /home/wm/Downloads/GameHub/data/com.github.tkashkin.gamehub.gschema.xml to /usr/share/glib-2.0/schemas
Installing /home/wm/Downloads/GameHub/build/data/com.github.tkashkin.gamehub.appdata.xml to /usr/share/metainfo
Installing /home/wm/Downloads/GameHub/build/data/com.github.tkashkin.gamehub.policy to /usr/share/polkit-1/actions
Installing /home/wm/Downloads/GameHub/data/com.github.tkashkin.gamehub-overlayfs-helper to /usr/bin
Running custom install script '/usr/bin/meson --internal gettext install --subdir=po --localedir=share/locale --pkgname=com.github.tkashkin.gamehub'
Installing /home/wm/Downloads/GameHub/build/po/ru.gmo to /usr/share/locale/ru/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/pt_BR.gmo to /usr/share/locale/pt_BR/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/id.gmo to /usr/share/locale/id/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/nb_NO.gmo to /usr/share/locale/nb_NO/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/nl.gmo to /usr/share/locale/nl/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/de.gmo to /usr/share/locale/de/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/fr.gmo to /usr/share/locale/fr/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/pl.gmo to /usr/share/locale/pl/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/tr.gmo to /usr/share/locale/tr/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/zh_CMN-HANT.gmo to /usr/share/locale/zh_CMN-HANT/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/el.gmo to /usr/share/locale/el/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/fa.gmo to /usr/share/locale/fa/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/hi.gmo to /usr/share/locale/hi/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/te.gmo to /usr/share/locale/te/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Installing /home/wm/Downloads/GameHub/build/po/ar.gmo to /usr/share/locale/ar/LC_MESSAGES/com.github.tkashkin.gamehub.mo
Running custom install script '/usr/bin/python /home/wm/Downloads/GameHub/meson/post_install.py'
Updating icon cache...
Compiling gsettings schemas...
Updating desktop database...
[wm@w-home build]$ gamehub --version
- GameHub
    Version: 0.14.2-23c164d-itch
    Branch:  itch
    Commit:  23c164d70d0f9d202654df4738ebbfe118214551
- Environment
    Distro:  EndeavourOS Linux
    DE:      i3
    GTK:     3.24.13
[wm@w-home build]$ gamehub
[INFO]   - GameHub
[INFO]       Version: 0.14.2-23c164d-itch
[INFO]       Branch:  itch
[INFO]       Commit:  23c164d70d0f9d202654df4738ebbfe118214551
[INFO]   - Environment
[INFO]       Distro:  EndeavourOS Linux
[INFO]       DE:      i3
[INFO]       GTK:     3.24.13
[INFO]       Themes:  Adwaita | Adwaita
Segmentation fault (core dumped)
[wm@w-home build]$ 
`

Should I try it on a fresh system that hasn't seen gamehub yet?

**Comment by hagabaka** [comment]
Thanks for helping with testing.

I also get a segfault if `butler` isn't installed. hagabaka@af359d9a362790accd661d508d33c95d0ba8ec44 fixes this. In the mean time you can also try if installing butler (available on AUR) works.

**Comment by spinningD20** [comment]
https://gist.github.com/spinningD20/db4a7da0472a68c08c21c6a27bc7e96e

I am beyond the segfault now, thanks for the pointer about butler.  

Observations so far:  
- clicking on the itch.io "authentication is required" in the list of services when gamehub initially opens, nothing happens.
- I navigated to the options, found the itch.io section, clicked "Generate Key" and went through that process.  
- Copy pasted the api key into the itch.io settings.  status of service changed to Authenticated, and I saw my itch.io titles show up in the list of games.
- Right clicked "Baba Is You" (great game) and pressed Install.  Install popup appeared, picked linux version, pushed Install.
- Baba Is You tile text displays "Starting Download", and noticed `[FATAL]  [Json] json_object_get_int_member: assertion 'node != NULL' failed` in the terminal log.  Nothing happens beyond this point.

**Comment by spinningD20** [comment]
I noticed a few different flags for running gamehub, and got it to give me some DEBUG level logging in the terminal.  after the FATAL json null error, I got this json response:

https://gist.github.com/spinningD20/0411ec16bd1751f15dcf38d9b041a63f

The only thing I noticed that was "null" is the `build` value?  Or perhaps this is some other response, I am not sure.  Figured I would mention it!

**Comment by hagabaka** [comment]
I reverted the commit which added game updates, but unfortunately also causes this problem sometimes. Can you try if the latest commit works?

**Comment by spinningD20** [comment]
we're getting further!  :)  

With the latest commit, I am able to download and install Baba Is You.  I went and installed Celeste as well.  

So with the happy path out of the way, I started going down some other scenarios...  one of them is canceling the download after it is started.  When I do this, it does technically stop the Download, but the UI state of the tile text is not updated, still shows me the (last update of) the download progress.  If I close gamehub and reopen it, it correctly shows not installed, and will let me go install it.

The only other side-effect of starting the download, canceling, closing gamehub, and reopening and reinstalling, is that it won't clean up the canceled download:  

![image](https://user-images.githubusercontent.com/3946418/71323166-36bcd880-249e-11ea-905e-a7e37b3fb232.png)

Not sure what the priority is, if we want any of the above fixed before it is merged in or not?  Testing this same scenario with GOG games, it uses a file in the GOG collection directory to store the downloaded data, and when canceled and re-started, it uses the same file (canceling the download does not make gamehub delete the progress it made in a previous download attempt).

More just letting you guys know this detail I noticed.  Otherwise, it seems to work as intended!  

(disclaimer, I've only installed/uninstalled/downloaded/cancelled download, I am sure there are other things to test too :)

**Comment by hagabaka** [comment]
With the latest commit on my branch, status should be updated correctly when cancelling an installation.

For clearing or reusing files from cancelled installations, I think there should be an easier way to let butler handle it, but I'm still trying to get the information from the Itch developer. I feel this could be done in a separate issue after the main features have been merged.

**Comment by spinningD20** [comment]
Thanks @hagabaka !  While I only have a few games to test with from itch.io (most purchases I've done through them are for art assets), I would say it's ready to merge, and we'll tackle more undiscovered, involved issues as they arise.

@tkashkin or hagabaka, just let me know if you'd like me to do anything further, if there's anything more I can do to help out.

**Comment by tkashkin** [comment]
I have merged it into `dev` branch.
It may need more tests covering various cases (with/without butler installed, authenticated/not authenticated, free/paid games, etc.).

**Comment by barfin** [comment]
> I have merged it into `dev` branch.
> It may need more tests covering various cases (with/without butler installed, authenticated/not authenticated, free/paid games, etc.).

clicking on itch io does not open up the login page

**Comment by tkashkin** [comment]
There is no login page, you need to open Settings and set your itch.io API key.

**Comment by barfin** [comment]
> @barfin There is no login page, you need to open Settings and set your itch.io API key.

oh i see now it worked

**Comment by hagabaka** [comment]
Is it possible to make the login page open the itch.io settings page, or show an information dialog? I think we can do that in `Itch.authenticate()` when the API key is empty or null, but I'm not sure how to open the settings page or show a dialog.

**Comment by 5310** [comment]
Thank you for Itch support! Linux natives games seem to run splendidly, at least all the ones I've tested so far, free or paid. 

But Windows games don't seem to launch at all. I don't even get an "executable cannot be found message", and even after setting it manually I get no "running with compatibility layer" text like the other Proton games. Does the Itch service no have any support for this yet?

**Comment by tkashkin** [comment]
@5310 Yes, compatibility layers are not supported for itch games yet. GameHub does not run games directly and butler does not expose available game executables.

**Comment by onnyyonn** [comment]
As @spinningD20 has already pointed out in his [comment](https://github.com/tkashkin/GameHub/issues/65#issuecomment-568192890), clicking on "authentication is required" in the list of services does nothing. The user has to generate an api key, the link for which is buried in the settings. This is confusing to an user who doesn't already know what to do. Can the behaviour of "authentication is required" tab be changed to that of the "generate key" link? Or opening a window explaining the steps may also be useful.

**Comment by captn3m0** [comment]
>Copy pasted the api key into the itch.io settings. status of service changed to Authenticated, and I saw my itch.io titles show up in the list of games.

This doesn't seem to work for me. I've pasted the generated key, I get a notice "Some settings will be applied after application restart".

I restart GameHub, but the itch.io status still says `Not Installed`. 

I've tried twice with different keys

**Comment by onnyyonn** [comment]
@captn3m0 I suspect (but not completely sure) that the Itch client (or Butler) has to be installed for Gamehub to move from `Not Installed` to `Authentication Required` and then it can access Itch via the key.

**Comment by captn3m0** [comment]
That was it! Seems to have fixed it for now, thanks.

It ought to be able to just run off my generated key though?

**Comment by RPGHank** [comment]
Sorry but I got lost somewhere in the thread
Is this coming to the official client or is it in a fork?

**Comment by RPGHank** [comment]
Since the itch client is open source is it possible to have itch support built in to Gamehub without having to install the Itch client?
Or is there something which makees this not possible?

**Comment by PeacefulPotato** [comment]
It still says "not installed" after pasting the key and installing both itch and butler. Butler is soft-linked to /usr/bin and /usr/bin/local and itch was installed normally.

**Comment by librarianmage** [comment]
May I ask what the status of this is?

**Comment by ghost** [comment]
> May I ask what the status of this is?

Implemented and i believe it should be closed;
people who are experiencing issues when using itch io in gamehub should open new issues
